prompb/io/prometheus/write/v2/types.proto

```diff
@@ -11,58 +11,55 @@
+// "application/x-protobuf;proto=io.prometheus.write.v2.Request"
+//
+// NOTE: gogoproto options might change in future for this file, they
+// are not part of the spec proto (they only modify the generated Go code, not
+// the serialized message). See: https://github.com/prometheus/prometheus/issues/11908
+message Request {
+  // symbols contains a de-duplicated array of string elements used for various
+  // items in a Request message, like labels and metadata items. To decode
   // each of those items, referenced, by "ref(s)" suffix, you need to lookup the
   // actual string by index from symbols array. The order of strings is up to
   // the client, server should not assume any particular encoding.
   repeated string symbols = 1;
-  // timeseries represents array of distinct series with 0 or more samples.
+  // timeseries represents an array of distinct series with 0 or more samples.
   repeated TimeSeries timeseries = 2 [(gogoproto.nullable) = false];
 }
 
 // TimeSeries represents a single series.
 message TimeSeries {
   // labels_refs is a list of label name-value pair references, encoded
-  // as indices to the WriteRequest.symbols array. This list's len is always
-  // a multiple of 2, and the underlying labels should be sorted.
+  // as indices to the Request.symbols array. This list's length is always
+  // a multiple of two, and the underlying labels should be sorted.
   //
   // Note that there might be multiple TimeSeries objects in the same
-  // WriteRequests with the same labels e.g. for different exemplars, metadata
+  // Requests with the same labels e.g. for different exemplars, metadata
   // or created timestamp.
   repeated uint32 labels_refs = 1;
 
-  // samples contain zero or more samples for a given timeseries. For typical
-  // clients, in healthy cases, there will be only one sample, for ~real
-  // time metric streaming. Samples can, in theory, co-exist with histogram samples
-  // (histograms field), although it should be extremely rare in practice (e.g.
-  // only when classic histogram series and native histogram share exactly the
-  // same metric name).
+  // Timeseries messages can either specify samples or (native) histogram samples
+  // (histogram field), but not both. For typical clients (~real-time metric
+  // streaming), in healthy cases, there will be only one sample or histogram.
   //
-  // Samples are sorted by timestamp (older first).
+  // Samples and histograms are sorted by timestamp (older first).
   repeated Sample samples = 2 [(gogoproto.nullable) = false];
-  // histograms contain zero or more histogram samples for a given timeseries.
-  // For typical clients, in healthy cases, there will be only one sample, for ~real
-  // time metric streaming. histograms can co-exist with samples (see samples
-  // for details).
-  //
-  // histograms are sorted by timestamp (older first).
   repeated Histogram histograms = 3 [(gogoproto.nullable) = false];
```

Comment:
Changes to samples & histogram samples behaviour I mentioned.

Reply:
Relevant discussion https://docs.google.com/document/d/1PljkX3YLLT-4f7MqrLt7XCVPG3IsjRREzYrUzBxCPV0/edit?disco=AAAA3dp46Gw